cvmfs_swissknife ingestsql

vvolkl opened the pull request.

vvolkl committed:
add dependency on libacl

vvolkl committed:
make IsSubPath available for ingestsql

vvolkl committed:
add ingestsql sources as is

vvolkl committed:
[todo] workaround for visibility

vvolkl committed:
add missing include

vvolkl committed:
fix dirent.mode_ -> dirent.mode(), log format

vvolkl committed:
more mode_ -> dirent.mode()

vvolkl committed:
add ingestsql friend

vvolkl committed:
Fix warning: (LIBACL)  does not match the name of the calling package (LibACL).

andrey-utkin commented:
Hello!
Which PR is going to be further worked on, this one or the earlier https://github.com/cvmfs/cvmfs/pull/3777 ?
Both of them fail for me in different ways.
I will gladly submit further patches for your branch if this PR is up to date.

This is a build failure log from Fedora 41:
[err.txt](https://github.com/user-attachments/files/19370195/err.txt)
Debian 12 failure log:
[deb12-err.txt](https://github.com/user-attachments/files/19371374/deb12-err.txt)

vvolkl commented:
Yes, this one will be merged, but  requires a bit more cleanup still. Let me quickly add the remaining fixes, then you can rebase here.

vvolkl committed:
up

vvolkl committed:
up

vvolkl committed:
codespell

vvolkl committed:
use syncparameters

andrey-utkin committed:
Implement replacement for libacl acl_to_xattr()

Up to now, the upcoming swissknife_ingestsql tool uses libacl (from
"acl" project) to convert ACL metadata from textual to xattr form.

But libacl does not export appropriate function.
For now a workaround of patching a bundled libacl was employed. The
functionality works, but dependency bundling can create friction to
adoption by some Linux distributions.

The needed patch was sent to the mailing list of acl package, but in the
last 1.5 months, it hasn't received any reaction.
https://lists.nongnu.org/archive/html/acl-devel/2025-04/msg00000.html

Exactly the same technical change was considered in 2018, but nothing
was changed: https://mail.gnu.org/archive/html/acl-devel/2018-02/msg00004.html

This changeset does not remove libacl dependency from the build system,
this will have to be done in a separate commit.

This changeset includes both the implementation and the tests.
Tests are based on reference output which is incorporated into the test
source code. For most of the test cases, the new implementation returns
exactly the same binary output and return code as the old one.

The new unittests module has a compile time option (disabled by default)
to run both new implementation and libacl functions on the test cases
and to compare their results. To enable this:

- uncomment `#define COMPARE_TO_LIBACL` in cvmfs/acl.h
- add `${LIBACL_LIBRARY}` to CVMFS_UNITTESTS_LINK_LIBRARIES in test/unittests/CMakeLists.txt

vvolkl committed:
cvmfs_config.h is deprecated

vvolkl committed:
fix warning

vvolkl committed:
codespell

vvolkl committed:
fix for linter workflow

vvolkl committed:
fix debian acl dependencies

vvolkl committed:
up

vvolkl commented on CMakeLists.txt:
```diff
@@ -35,6 +35,7 @@ endif()
 #
 set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/cmake/Modules/")
 
+
```
```suggestion
```
whitespace

andrey-utkin commented on .github/workflows/linter.yaml:
```diff
@@ -38,6 +38,7 @@ jobs:
       - name: Install build dependencies
         run: |
           sudo apt-get install -y /tmp/cvmfs-build-deps_2.1.0_all.deb
+          sudo apt-get install -y libacl1-dev #TODO: figure out why not installed with mk-build-deps
```
@vvolkl  `COMPARE_TO_LIBACL` checking functionality still requires *patched* libacl. So not sure why you are adding this dependency, I am not aware of any usage of libacl beyond `marshal_xattrs()`.

vvolkl replied:
Right, that's  true. I'll just remove it from the build system then. I'll set up a separate test environment for those tests.

vvolkl committed:
whitespace

vvolkl committed:
remove libacl from build deps

vvolkl commented on cvmfs/CMakeLists.txt:
```diff
@@ -699,6 +702,7 @@ if (BUILD_SERVER)
         ${VJSON_LIBRARIES}
         ${CAP_LIBRARIES}
         ${LibArchive_LIBRARY}
+        ${LIBACL_LIBRARY}
```
```suggestion
```

vvolkl committed:
Update cvmfs/CMakeLists.txt

vvolkl commented:
@andrey-utkin The unittests seem to fail on all debian based platforms:
```
[----------] 7 tests from T_Acl
[ RUN      ] T_Acl.t2
[       OK ] T_Acl.t2 (0 ms)
[ RUN      ] T_Acl.t1
[       OK ] T_Acl.t1 (0 ms)
[ RUN      ] T_Acl.t8
[       OK ] T_Acl.t8 (0 ms)
[ RUN      ] T_Acl.t3
/home/sftnight/jenkins/workspace/CvmfsUnittest/CVMFS_BUILD_ARCH/docker-x86_64/CVMFS_BUILD_PLATFORM/debian12/cvmfs/test/unittests/t_acl.cc:57: Failure
      Expected: 0
To be equal to: memcmp(binary_acl, acl_binary_expected, binary_size)
      Which is: 1
[  FAILED  ] T_Acl.t3 (0 ms)
[ RUN      ] T_Acl.t5
/home/sftnight/jenkins/workspace/CvmfsUnittest/CVMFS_BUILD_ARCH/docker-x86_64/CVMFS_BUILD_PLATFORM/debian12/cvmfs/test/unittests/t_acl.cc:57: Failure
      Expected: 0
To be equal to: memcmp(binary_acl, acl_binary_expected, binary_size)
      Which is: 1
[  FAILED  ] T_Acl.t5 (0 ms)
[ RUN      ] T_Acl.t9
/home/sftnight/jenkins/workspace/CvmfsUnittest/CVMFS_BUILD_ARCH/docker-x86_64/CVMFS_BUILD_PLATFORM/debian12/cvmfs/test/unittests/t_acl.cc:57: Failure
      Expected: 0
To be equal to: memcmp(binary_acl, acl_binary_expected, binary_size)
      Which is: -1
[  FAILED  ] T_Acl.t9 (0 ms)
[ RUN      ] T_Acl.t6
/home/sftnight/jenkins/workspace/CvmfsUnittest/CVMFS_BUILD_ARCH/docker-x86_64/CVMFS_BUILD_PLATFORM/debian12/cvmfs/test/unittests/t_acl.cc:57: Failure
      Expected: 0
To be equal to: memcmp(binary_acl, acl_binary_expected, binary_size)
      Which is: -2
[  FAILED  ] T_Acl.t6 (0 ms)
[----------] 7 tests from T_Acl (1 ms total)

```

andrey-utkin commented:
> The unittests seem to fail on all debian based platforms:

Sorry for that!
I looked at this yesterday, turns out not all basic groups are the same across distros - `bin` group is 2 on Debian, but 1 on Fedora and Gentoo. I will replace it groups which have the same GID across distros: `root` (0), `sys` (3), `tty` (5).

I will submit the linting fixes for my patch as well.

andrey-utkin committed:
[test] fix ACL tests failures on Debian

This was due to mismatch of GID and UID of entities present in all
distros: bin, daemon.

andrey-utkin committed:
acl: linting fixes

andrey-utkin commented:
@vvolkl please pull https://github.com/vvolkl/cvmfs/pull/5

vvolkl commented:
Great, thanks a lot for the quick fix!

andrey-utkin commented:
@vvolkl am I understanding the CI issues correctly?
- cc9/aarch64, ubuntu2404/aarch64 complaining about go version "1.23" vs "1.23.0" in go.mod, perhaps because they have go older than 1.23 (cached? package-manager-provided?)
- ubuntu2404/x86_64 fails to compile for the lack of [my patch](https://github.com/cvmfs/cvmfs/pull/3793/commits/30f226aa654fe8c8e23d6dcf06a3f4ce6dc86dcf) - maybe we could `ifdef` our way out of the issue with support of older distros?
- linter: `no template named 'CatalogMergeTool'` - I don't see it with `clang-tidy cvmfs/receiver/catalog_merge_tool_impl.h` with Clang 19, going to run with 18.

I will also look how to contribute to tests and documentation.
Are there any other blockers?

vvolkl committed:
fix includes

vvolkl committed:
clean up more comments

vvolkl commented:
> @vvolkl am I understanding the CI issues correctly?
> 
>     * cc9/aarch64, ubuntu2404/aarch64 complaining about go version "1.23" vs "1.23.0" in go.mod, perhaps because they have go older than 1.23 (cached? package-manager-provided?)

That's a caching issue. Some of the test nodes do incremental builds and hit an issue with a recent update of golang, can be ignored.

> 
>     * ubuntu2404/x86_64 fails to compile for the lack of [my patch](https://github.com/cvmfs/cvmfs/pull/3793/commits/30f226aa654fe8c8e23d6dcf06a3f4ce6dc86dcf) - maybe we could `ifdef` our way out of the issue with support of older distros?

Rebased on devel now, I think that should be fixed with just adding the include


> 
>     * linter: `no template named 'CatalogMergeTool'` - I don't see it with `clang-tidy cvmfs/receiver/catalog_merge_tool_impl.h` with Clang 19, going to run with 18.
> 

That's a false positive for sure. I'm anyway updating clang-tidy to 19 in #3853 
Clang tidy did suggest a number of good fixes, if you can fix those that would be great! (see attachment)
[clang-tidy-ingestsql.txt](https://github.com/user-attachments/files/20551443/clang-tidy-ingestsql.txt)


> 
> I will also look how to contribute to tests and documentation. Are there any other blockers?

No blockers. Some more refactoring and tests are needed for sure, but we can do that in follow-up PRs.

vvolkl committed:
Merge branch 'devel' into ingestsql

vvolkl commented:
There's some more work needed for test coverage and refactoring, but I'll merge to make collaborating on this feature easier.
